components/salsa-2022/src/runtime/local_state.rs

```diff
@@ -80,19 +75,23 @@ pub enum QueryOrigin {
 }
 
 impl QueryOrigin {
-    /// Indices for queries *written* by this query (or `&[]` if its value was assigned).
-    pub(crate) fn outputs(&self) -> impl Iterator<Item = DependencyIndex> + '_ {
-        let slice = match self {
+    /// Indices for queries *written* by this query (or `vec![]` if its value was assigned).
+    pub(crate) fn outputs(&self) -> Box<dyn Iterator<Item = DependencyIndex> + '_> {
```

Comment:
Nothing *wrong* with returning a `Box<dyn>`, but in this kind of scenario I usually use ye olde flat-map trick...

```rust
pub(crate) fn outputs(&self) -> impl Iterator<Item = DependencyIndex> + '_ {
    let opt_edges = match self {
        QueryOrigin::Derived(edges) | QueryOrigin::DerivedUntracked(edges) => Some(edges),
        QueryOrigin::Assigned(_) | QueryOrigin::BaseInput => None,
    };
    opt_edges.flat_map(|edges| edges.outputs())
}
```

Reply:
It works well when you have one "iterator" case and one "empty" case.

Reply:
This is exactly what I was looking for, thanks! @nikomatsakis